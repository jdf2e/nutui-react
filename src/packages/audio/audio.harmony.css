.nut-audio-icon {
  position: relative;
  display: inline-block;
}
.nut-audio-icon-box {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 30px;
  height: 30px;
  background: #ffffff;
  border-radius: 50%;
  box-shadow: 0 0 8px #bfbfbf;
}
.nut-audio-icon .nut-audio-icon-stop {
  position: relative;
}
.nut-audio-icon .nut-audio-icon-stop::after {
  position: absolute;
  left: 50%;
  top: 50%;
  transform: translateX(-50%);
  content: "";
  height: 2px;
  width: 30px;
  background: #bfbfbf;
  transform: rotate(45deg);
  transform-origin: 8px -18px;
}
.nut-audio-progress {
  display: flex;
  align-items: center;
  width: 100%;
  margin: 0px auto;
  padding: 10px 0;
}
.nut-audio-progress-bar-wrapper {
  flex: 1;
  margin: 0 10px;
}
.nut-audio-progress .time {
  min-width: 50px;
  font-size: 12px;
  text-align: center;
}
.nut-audio-progress .nut-range-button {
  width: 8px;
  height: 8px;
}
.nut-audio .custom-button-group .nut-button-primary {
  margin: 0 5px;
}
.nut-audio .custom-button-group-disable .nut-button-primary {
  margin: 0 5px;
  pointer-events: none;
}
.nut-audio .disable {
  color: blue;
}
.nut-audio .nut-audio-none-container .nut-voice {
  border: 1px solid #1a1a1a;
  align-items: center;
}

[dir=rtl] .nut-audio-icon .nut-audio-icon-stop::after,
.nut-rtl .nut-audio-icon .nut-audio-icon-stop::after {
  left: auto;
  right: 50%;
  transform: rotate(-45deg);
  transform-origin: 20px -18px;
}