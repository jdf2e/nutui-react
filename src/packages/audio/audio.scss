@import '../icon/icon.scss';
@import '../button/button.scss';
@import '../range/range.scss';
.nut-theme-dark {
  .nut-audio__icon {
    .nut-audio__icon-box {
      background: $dark-color;
    }
  }
}
.nut-audio {
  &__icon {
    position: relative;
    display: inline-block;
    &-box {
      display: flex;
      align-items: center;
      justify-content: center;
      width: 30px;
      height: 30px;
      background: #fff;
      border-radius: 50%;
      box-shadow: 0 0 8px rgba($color: $text-color, $alpha: 0.5);

      .nut-icon-am-rotate {
        transform: rotate((360));
        animation-timing-function: ease-in-out;
      }
    }

    .nut-audio__icon-stop {
      position: relative;
      &::after {
        position: absolute;
        left: 50%;
        top: 50%;
        transform: translateX(-50%);
        content: '';
        height: 2px;
        width: 30px;
        background: $disable-color;
        transform: rotate(45deg);
        transform-origin: 8px -18px;
      }
    }
  }

  &__progress {
    display: flex;
    align-items: center;
    width: 100%;
    margin: 0px auto;
    padding: 10px 0;

    &-bar-wrapper {
      flex: 1;
      margin: 0 10px;
    }
    .time {
      min-width: 50px;
      font-size: 12px;
      text-align: center;
    }
    .nut-range-button {
      width: 8px;
      height: 8px;
    }
  }
  .custom-button-group {
    .nut-button--primary {
      margin: 0 5px;
    }
  }
  .custom-button-group-disable {
    .nut-button--primary {
      margin: 0 5px;
      pointer-events: none;
    }
  }
  .disable {
    color: blue;
  }
}
