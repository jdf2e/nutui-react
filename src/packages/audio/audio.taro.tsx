import React, {
  useState,
  useRef,
  useEffect,
  CSSProperties,
  FunctionComponent,
} from 'react'

import Taro from '@tarojs/taro'
import Icon from '@/packages/icon/index.taro'
import Range from '@/packages/range/index.taro'
import Button from '@/packages/button/index.taro'
import bem from '@/utils/bem'
import { useConfig } from '@/packages/configprovider/configprovider.taro'
import { BasicComponent, ComponentDefaults } from '@/utils/typings'

const b = bem('audio')
const warn = console.warn

function formatSeconds(value: string): string {
  if (!value) {
    return '00:00:00'
  }
  const time = parseInt(value)
  const hours = Math.floor(time / 3600)
  const minutes = Math.floor((time - hours * 3600) / 60)
  const seconds = time - hours * 3600 - minutes * 60
  let result = ''
  result += `${`0${hours.toString()}`.slice(-2)}:`
  result += `${`0${minutes.toString()}`.slice(-2)}:`
  result += `0${seconds.toString()}`.slice(-2)
  return result
}

export interface AudioProps extends BasicComponent {
  className?: string
  style?: CSSProperties
  url: string
  autoplay?: boolean
  loop?: boolean
  type: string
  onFastBack?: (ctx: Taro.InnerAudioContext) => void
  onForward?: (ctx: Taro.InnerAudioContext) => void
  onPause?: any
  onPlay?: any
  onPlayEnd?: (ctx: Taro.InnerAudioContext) => void
  onCanPlay?: (ctx: Taro.InnerAudioContext) => void
}
const defaultProps = {
  ...ComponentDefaults,
  className: '',
  url: '',
  style: {},
  autoplay: false,
  loop: false,
  type: 'progress',
  onFastBack: (ctx: Taro.InnerAudioContext) => {}, // type 为 progress时生效
  onForward: (ctx: Taro.InnerAudioContext) => {}, // type 为 progress时生效
  onPause: (ctx: Taro.InnerAudioContext) => {},
  onPlay: (ctx: Taro.InnerAudioContext) => {},
  onPlayEnd: (ctx: Taro.InnerAudioContext) => {},
  onCanPlay: (ctx: Taro.InnerAudioContext) => {},
} as AudioProps
export const Audio: FunctionComponent<
  Partial<AudioProps> &
    (React.HTMLAttributes<HTMLDivElement> | Taro.InnerAudioContext)
> = (props) => {
  const { locale } = useConfig()
  const {
    className,
    url,
    style,
    autoplay,
    loop,
    type,
    onFastBack,
    onForward,
    onPause,
    onPlay,
    onPlayEnd,
    onCanPlay,
    children,
    iconClassPrefix,
    iconFontClassName,
    ...rest
  } = {
    ...defaultProps,
    ...props,
  } as any

  const [playing, setPlaying] = useState(false)
  const [totalSeconds, setTotalSeconds] = useState(0)
  const [percent, setPercent] = useState(0)
  const [isCanPlay, setIsCanPlay] = useState(false)
  const [currentDuration, setCurrentDuration] = useState('00:00:00')

  const statusRef = useRef({
    currentTime: 0,
    currentDuration: '00:00:00',
    percent: 0,
  })

  const audioRef = useRef(Taro.createInnerAudioContext())
  const audioCtx = audioRef.current

  useEffect(() => {
    // initial
    statusRef.current = {
      currentTime: 0,
      currentDuration: '00:00:00',
      percent: 0,
    }
    setIsCanPlay(false)
    setTotalSeconds(0)
    setPercent(0)
    setCurrentDuration('00:00:00')
    setPlaying(false)

    audioCtx.src = url
    audioCtx.autoplay = autoplay || false
    audioCtx.loop = loop || false
    audioCtx.onPause(() => {
      props.onPause && props.onPause(audioCtx)
    })
    audioCtx.onEnded(() => {
      audioCtx.seek(0)
      setPlaying(false)
      setPercent(0)
      setCurrentDuration(formatSeconds('0'))
      if (props.loop) {
        warn(locale.audio.tips || 'onPlayEnd事件在loop=false时才会触发')
      } else {
        props.onPlayEnd && props.onPlayEnd(audioCtx)
      }
    })

    audioCtx.onPlay(() => {
      const { duration } = audioCtx
      setTotalSeconds(Math.floor(duration))
      props.onPlay && props.onPlay(audioCtx)
    })
    audioCtx.onCanplay(() => {
      // hack: 在 onCanPlay 读取一次 paused 可以获取最新的 audioCtx 状态
      const isPaused = audioCtx.paused
      setPlaying(!isPaused)
      const intervalID = setInterval(function () {
        if (audioCtx.duration !== 0) {
          setTotalSeconds(audioCtx.duration)
          clearInterval(intervalID)
        }
      }, 500)
      setIsCanPlay(true)
      props.onCanPlay && props.onCanPlay(audioCtx)
    })
    audioCtx.onTimeUpdate(() => {
      const time = parseInt(`${audioCtx.currentTime}`)
      const formated = formatSeconds(`${time}`)
      statusRef.current.currentDuration = formated
      setPercent((time / totalSeconds) * 100)
      setCurrentDuration(formatSeconds(audioCtx.currentTime.toString()))
    })

    audioCtx.onStop(() => {
      setPlaying(false)
      setCurrentDuration(formatSeconds('0'))
      setPercent(0)
      audioCtx.seek(0)
    })

    audioCtx.onError((res) => {
      console.log('code', res.errCode)
      console.log('message', res.errMsg)
    })
  }, [audioCtx, autoplay, url, loop])

  useEffect(() => {
    return () => {
      audioCtx.destroy()
    }
  }, [])

  const handleBack = () => {
    const currentTime = Math.floor(audioCtx.currentTime)
    statusRef.current.currentTime = Math.max(currentTime - 1, 0)
    setPercent((statusRef.current.currentTime / totalSeconds) * 100)
    setCurrentDuration(formatSeconds(statusRef.current.currentTime.toString()))
    audioCtx.seek(statusRef.current.currentTime)
    props.onFastBack && props.onFastBack(audioCtx)
  }

  const handleForward = () => {
    const currentTime = Math.floor(audioCtx.currentTime)
    statusRef.current.currentTime = Math.min(currentTime + 1, audioCtx.duration)
    setPercent((statusRef.current.currentTime / totalSeconds) * 100)
    setCurrentDuration(formatSeconds(statusRef.current.currentTime.toString()))
    audioCtx.seek(statusRef.current.currentTime)
    props.onForward && props.onForward(audioCtx)
  }

  const handleStatusChange = () => {
    setPlaying(!playing)
    if (!playing) {
      audioCtx.play()
    } else {
      audioCtx.pause()
    }
  }

  const renderIcon = () => {
    return (
      <>
        <div className={b('icon')}>
          <div
            className={`${b('icon-box')} ${
              playing ? b('icon-play') : b('icon-stop')
            }`}
            onClick={handleStatusChange}
          >
            {playing ? (
              <Icon
                classPrefix={iconClassPrefix}
                fontClassName={iconFontClassName}
                name="service"
                className="nut-icon-loading"
              />
            ) : (
              <Icon
                classPrefix={iconClassPrefix}
                fontClassName={iconFontClassName}
                name="service"
              />
            )}
          </div>
        </div>
      </>
    )
  }

  const renderProgerss = () => {
    return (
      <>
        <div className={b('progress')}>
          <div className="time">{currentDuration}</div>
          <div className={b('progress-bar-wrapper')}>
            <Range
              modelValue={percent}
              hiddenTag
              hiddenRange
              inactive-color="#cccccc"
              active-color="#fa2c19"
            />
          </div>
          <div className="time">
            {formatSeconds(`${totalSeconds}`) || '00:00:00'}
          </div>
        </div>
        <div
          className={
            isCanPlay ? 'custom-button-group' : 'custom-button-group-disable'
          }
        >
          <Button
            type="primary"
            size="small"
            className="back"
            onClick={handleBack}
          >
            {locale.audio.back || '快退'}
          </Button>
          <Button
            type="primary"
            size="small"
            className="start"
            onClick={handleStatusChange}
          >
            {playing
              ? `${locale.audio.pause || '暂停'}`
              : `${locale.audio.start || '开始'}`}
          </Button>
          <Button type="primary" size="small" onClick={handleForward}>
            {locale.audio.forward || '快进'}
          </Button>
        </div>
      </>
    )
  }

  const renderNone = () => {
    return (
      <div className={b('none-container')} onClick={handleStatusChange}>
        {children}
      </div>
    )
  }

  const renderAudio = () => {
    switch (type) {
      case 'icon':
        return renderIcon()
      case 'progress':
        return renderProgerss()
      case 'none':
        return renderNone()
      default:
        return null
    }
  }

  return (
    <div className={`${b()} ${className}`} style={style} {...rest}>
      {renderAudio()}
    </div>
  )
}

Audio.defaultProps = defaultProps
Audio.displayName = 'NutAudio'
