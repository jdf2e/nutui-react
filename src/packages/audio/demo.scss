.audio-demo {
  padding-top: 57px;
  -webkit-box-direction: normal;
  -webkit-tap-highlight-color: transparent;
  box-sizing: border-box;
  margin: 0;
  border: 0;
  font-family: PingFang SC, Microsoft YaHei, Helvetica, Hiragino Sans GB, SimSun,
    sans-serif;
  height: 100%;
  background: #f7f8fa;
  overflow-x: hidden;
  overflow-y: auto;
  padding: 57px 17px 0;
  h2 {
    -webkit-box-direction: normal;
    -webkit-tap-highlight-color: transparent;
    box-sizing: border-box;
    margin: 0;
    border: 0;
    font-family: PingFang SC, Microsoft YaHei, Helvetica, Hiragino Sans GB,
      SimSun, sans-serif;
    margin-top: 30px;
    margin-bottom: 10px;
    font-size: 14px;
    color: #909ca4;
    padding: 0 10px;
    font-weight: 400;
  }
  .nut-voice {
    display: flex;
    justify-content: space-between;
    width: 100px;
    height: auto;
    padding: 8px;
    border: 1px solid rgba(0, 0, 0, 0.6);
    border-radius: 18px;
  }
}
