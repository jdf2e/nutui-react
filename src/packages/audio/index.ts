import { Audio } from './audio'

export type { AudioProps } from './audio'
export default Audio
