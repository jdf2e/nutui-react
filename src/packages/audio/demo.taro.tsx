import React, { useState } from 'react'
import { Icon, Cell, Audio } from '@/packages/nutui.react.taro'
import '@/packages/audio/demo.scss'

const formatseconds = (value: string) => {
  if (!value) {
    return '00:00:00'
  }
  const time = parseInt(value)
  const hours = Math.floor(time / 3600)
  const minutes = Math.floor((time - hours * 3600) / 60)
  const secondss = time - hours * 3600 - minutes * 60
  let result = ''
  result += `${`0${hours.toString()}`.slice(-2)}:`
  result += `${`0${minutes.toString()}`.slice(-2)}:`
  result += `0${secondss.toString()}`.slice(-2)
  return result
}

const AudioDemo = () => {
  const [duration, setDuration] = useState('0')
  const [voiceIcon, setVoiceIcon] = useState('play-circle-fill')
  return (
    <div className="audio-demo">
      <div>
        <h2>1.Taro原生的Audio组件,1.6.0版本开始，该组件不再维护</h2>
        <h2>
          2.本组件基于Taro.createInnerAudioContext做了封装,由于api的限制,相比于NuiUI-React
          Audio组件,部分属性和事件发生了改动:
          <h2>(1) 属性移除:muted | preload | type = controls </h2>
          <h2>(2) 事件移除:onMute</h2>
          <h2>(3) 事件新增:onPlay</h2>
        </h2>
      </div>

      <h2>type=icon</h2>
      <Cell>
        <Audio
          autoplay={false}
          url="https://storage.360buyimg.com/jdcdkh/SMB/VCG231024564.wav"
          type="icon"
          loop={false}
          onPlayEnd={() => console.log('ended!')}
        />
      </Cell>
      <h2>type=progress</h2>
      <Cell>
        <Audio
          autoplay={false}
          style={{ fontSize: '20px' }}
          url="https://storage.360buyimg.com/jdcdkh/SMB/VCG231024564.wav"
          type="progress"
          onForward={() => console.log('forward')}
          onPause={(ctx: any) => {
            console.log('progress audio paused', ctx)
          }}
          onPlayEnd={() => console.log('progress audio ended!')}
        />
      </Cell>
      <h2>type=none</h2>
      <Cell>
        <Audio
          className="custom-voice-audio"
          autoplay={false}
          url="https://storage.360buyimg.com/jdcdkh/SMB/VCG231024564.wav"
          type="none"
          onPlay={(ctx: any) => {
            setDuration(formatseconds(`${ctx.duration}`))
            setVoiceIcon('poweroff-circle-fill')
          }}
          onPause={(ctx: any) => {
            setVoiceIcon('play-circle-fill')
          }}
        >
          <div className="nut-voice">
            <Icon name={voiceIcon} />
            <div>{duration}&quot;</div>
          </div>
        </Audio>
      </Cell>
    </div>
  )
}
export default AudioDemo
