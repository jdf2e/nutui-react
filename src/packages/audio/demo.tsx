import React, { useState } from 'react'
import { Audio } from './audio'
import Icon from '../icon'
import './demo.scss'
import Cell from '../cell'

const AudioDemo = () => {
  const [duration, setDuration] = useState(0)
  const [voiceIcon, setVoiceIcon] = useState('play-circle-fill')
  const audioElement = document.querySelectorAll('audio')[2]
  if (audioElement) {
    audioElement.addEventListener('playing', (e) => {
      setVoiceIcon('poweroff-circle-fill')
    })
    audioElement.addEventListener('pause', (e) => {
      setVoiceIcon('play-circle-fill')
    })
  }
  return (
    <div className="demo">
      <h2>type=icon</h2>
      <Cell>
        <Audio
          autoplay={false}
          url="//storage.360buyimg.com/jdcdkh/SMB/VCG231024564.wav"
          type="icon"
          loop={false}
          preload="auto"
          muted={false}
          onPlayEnd={() => alert('ended!')}
        />
      </Cell>
      <h2>type=progress</h2>
      <Cell>
        <Audio
          autoplay={false}
          style={{ fontSize: '20px' }}
          url="//storage.360buyimg.com/jdcdkh/SMB/VCG231024564.wav"
          type="progress"
          preload="auto"
          muted={false}
          onMute={(e) => {
            console.log('progress audio muted', e)
          }}
          onForward={() => console.log('forward')}
          onPause={(e) => {
            console.log('progress audio paused', e)
          }}
          onPlayEnd={() => alert('progress audio ended!')}
        />
      </Cell>
      <h2>type=none</h2>
      <Cell>
        <Audio
          className="custom-voice-audio"
          id="custom-voice-audio"
          autoplay={false}
          url="//storage.360buyimg.com/jdcdkh/SMB/VCG231024564.wav"
          type="none"
          preload="auto"
          onCanPlay={(e: any) => {
            console.log('none-canplay', e)
            setDuration(e?.target?.duration.toFixed(0) || 0)
          }}
        >
          <div className="nut-voice">
            <Icon name={voiceIcon} />
            <div>{duration}&quot;</div>
          </div>
        </Audio>
      </Cell>
      <h2>type=controls</h2>
      <Cell>
        <Audio
          autoplay={false}
          url="//storage.360buyimg.com/jdcdkh/SMB/VCG231024564.wav"
          type="controls"
          preload="auto"
          muted={false}
          onPause={(e) => {
            console.log('paused', e)
          }}
          onPlayEnd={() => alert('ended!')}
        />
      </Cell>
    </div>
  )
}
export default AudioDemo
