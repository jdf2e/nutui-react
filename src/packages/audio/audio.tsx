import React, {
  useState,
  useEffect,
  useRef,
  FunctionComponent,
  ReactEventHandler,
} from 'react'
import { Service } from '@nutui/icons-react'
import classNames from 'classnames'
import Range from '@/packages/range'
import Button from '@/packages/button'
import { useConfig } from '@/packages/configprovider'

import { BasicComponent, ComponentDefaults } from '@/utils/typings'

export interface AudioProps extends BasicComponent {
  src: string
  muted: boolean
  autoplay: boolean
  loop: boolean
  preload: string
  type: string
  onBack: (e: HTMLAudioElement) => void
  onForward: (e: HTMLAudioElement) => void
  onPause: ((e: HTMLAudioElement) => void) & ReactEventHandler<HTMLAudioElement>
  onEnd: (e: HTMLAudioElement) => void
  onMute: (e: HTMLAudioElement) => void
  onCanPlay: ((e: HTMLAudioElement) => void) &
    ReactEventHandler<HTMLAudioElement>
}

const defaultProps = {
  ...ComponentDefaults,
  src: '',
  muted: false,
  autoplay: false,
  loop: false,
  preload: 'auto',
  type: 'progress',
  onBack: (e: HTMLAudioElement) => {}, // type 为 progress时生效
  onForward: (e: HTMLAudioElement) => {}, // type 为 progress时生效
  onPause: (e) => {},
  onEnd: (e: HTMLAudioElement) => {},
  onMute: (e: HTMLAudioElement) => {},
  onCanPlay: (e: HTMLAudioElement) => {},
} as AudioProps
export const Audio: FunctionComponent<
  Partial<AudioProps> & React.HTMLAttributes<HTMLElement>
> = (props) => {
  const { locale } = useConfig()
  const {
    className,
    src,
    style,
    muted,
    autoplay,
    loop,
    preload,
    type,
    onBack,
    onForward,
    onPause,
    onEnd,
    onMute,
    onCanPlay,
    children,
    ...rest
  } = {
    ...defaultProps,
    ...props,
  }

  const [playing, setPlaying] = useState(false)
  const [percent, setPercent] = useState(0)
  const [isCanPlay, setIsCanPlay] = useState(false)
  const [currentDuration, setCurrentDuration] = useState('00:00:00')

  const AudioRef = useRef<HTMLAudioElement>(null)
  const statusRef = useRef({
    currentTime: 0,
    currentDuration: '00:00:00',
    percent: 0,
    duration: '00:00:00',
    second: 0,
    hanMuted: props.muted,
    playing: props.autoplay,
    handPlaying: false,
  })
  const classPrefix = 'nut-audio'
  const warn = console.warn
  const handleEnded = (e: any) => {
    if (props.loop) {
      warn(locale.audio.tips || 'onPlayEnd事件在loop=false时才会触发')
    } else {
      props.onEnd && props.onEnd(e)
    }
  }

  function watch() {
    if (AudioRef && AudioRef.current) {
      const current = AudioRef.current
      current.addEventListener('play', () => {
        setPlaying(true)
      })
    }
  }
  useEffect(() => {
    watch()
  }, [])

  useEffect(() => {}, [currentDuration])
  const handleStatusChange = () => {
    setPlaying(!playing)
    if (playing) {
      AudioRef && AudioRef.current && AudioRef.current.pause()
    } else {
      AudioRef && AudioRef.current && AudioRef.current.play()
    }
  }
  const renderIcon = () => {
    return (
      <div className={`${classPrefix}__icon`}>
        <div
          className={classNames(
            `${classPrefix}__icon-box`,
            playing ? `${classPrefix}__icon-play` : `${classPrefix}__icon-stop`
          )}
          onClick={handleStatusChange}
        >
          <Service className={playing ? 'nut-icon-loading' : ''} />
        </div>
      </div>
    )
  }

  const handleBack = () => {
    if (statusRef.current.currentTime > 0 && AudioRef.current) {
      statusRef.current.currentTime--
      AudioRef.current.currentTime = statusRef.current.currentTime
      props.onBack && props.onBack(AudioRef.current)
    }
  }
  const handleForward = () => {
    if (AudioRef.current) {
      statusRef.current.currentTime++
      AudioRef.current.currentTime = statusRef.current.currentTime
      props.onForward && props.onForward(AudioRef.current)
    }
  }
  const handleMute = () => {
    if (AudioRef.current) {
      AudioRef.current.muted = !AudioRef.current.muted
      props.onMute && props.onMute(AudioRef.current)
    }
  }
  const handlePause = (e: HTMLAudioElement) => {
    setPlaying(false)
    props.onPause && props.onPause(e)
  }

  const formatSeconds = (value: string) => {
    if (!value) {
      return '00:00:00'
    }
    const time = parseInt(value)
    const hours = Math.floor(time / 3600)
    const minutes = Math.floor((time - hours * 3600) / 60)
    const seconds = time - hours * 3600 - minutes * 60
    let result = ''
    result += `${`0${hours.toString()}`.slice(-2)}:`
    result += `${`0${minutes.toString()}`.slice(-2)}:`
    result += `0${seconds.toString()}`.slice(-2)
    return result
  }
  const renderProgerss = () => {
    return (
      <>
        <div className={`${classPrefix}__progress`}>
          <div className="time">{currentDuration}</div>
          <div className={`${classPrefix}__progress-bar-wrapper`}>
            <Range
              modelValue={percent}
              hiddenTag
              hiddenRange
              inactive-color="#cccccc"
              active-color="#fa2c19"
            />
          </div>
          <div className="time">
            {AudioRef.current
              ? formatSeconds(`${statusRef.current.second}`)
              : '00:00:00'}
          </div>
        </div>
        <div
          className={
            isCanPlay ? 'custom-button-group' : 'custom-button-group-disable'
          }
        >
          <Button
            type="primary"
            size="small"
            className="back"
            onClick={handleBack}
          >
            {locale.audio.back || '快退'}
          </Button>
          <Button
            type="primary"
            size="small"
            className="start"
            onClick={handleStatusChange}
          >
            {playing
              ? `${locale.audio.pause || '暂停'}`
              : `${locale.audio.start || '开始'}`}
          </Button>
          <Button type="primary" size="small" onClick={handleForward}>
            {locale.audio.forward || '快进'}
          </Button>
          <Button
            type={
              AudioRef.current && AudioRef.current.muted ? 'default' : 'primary'
            }
            size="small"
            onClick={handleMute}
          >
            {locale.audio.mute || '静音'}
          </Button>
        </div>
      </>
    )
  }

  const renderNone = () => {
    return (
      <div
        className={`${classPrefix}__none-container`}
        onClick={handleStatusChange}
      >
        {children}
      </div>
    )
  }
  const renderAudio = () => {
    switch (type) {
      case 'icon':
        return renderIcon()
      case 'progress':
        return renderProgerss()
      case 'none':
        return renderNone()
      default:
        return null
    }
  }

  const handleCanplay = (e: any) => {
    setIsCanPlay(true)
    if (props.autoplay && !playing) {
      AudioRef && AudioRef.current && AudioRef.current.play()
    }
    if (AudioRef.current) {
      statusRef.current.second = AudioRef.current.duration || 0
      props.onCanPlay && props.onCanPlay(e)
    }
  }
  const onTimeupdate = (e: any) => {
    const time = parseInt(e.target.currentTime)
    const formated = formatSeconds(`${time}`)
    statusRef.current.currentDuration = formated
    setPercent((time / statusRef.current.second) * 100)
    setCurrentDuration(formated)
    statusRef.current.currentTime = time
  }
  return (
    <div className={classNames(classPrefix, className)} style={style} {...rest}>
      {renderAudio()}
      <audio
        className="audioMain"
        controls={type === 'controls'}
        ref={AudioRef}
        src={src}
        muted={muted}
        preload={preload}
        loop={loop}
        onPause={(e: any) => handlePause(e)}
        onEnded={(e: React.SyntheticEvent<HTMLAudioElement, Event>) =>
          handleEnded(e)
        }
        onCanPlay={(e) => handleCanplay(e)}
        onTimeUpdate={(e) => onTimeupdate(e)}
      >
        <track kind="captions" />
      </audio>
    </div>
  )
}

Audio.defaultProps = defaultProps
Audio.displayName = 'NutAudio'
