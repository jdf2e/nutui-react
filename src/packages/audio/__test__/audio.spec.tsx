import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'
import { Audio } from '../audio'

test('className && style && type = progress  && onPause  && onFastBack && onForward test ', () => {
  const playeEnd = jest.fn()
  const pause = jest.fn()
  const canPlay = jest.fn()
  const mute = jest.fn()
  const fastBack = jest.fn()
  const forward = jest.fn()
  const TestComponent = () => {
    return (
      <>
        <Audio
          className="audio-jest"
          style={{ fontSize: '20px' }}
          type="progress"
          autoplay
          src="//storage.360buyimg.com/jdcdkh/SMB/VCG231024564.wav"
          onPlayEnd={playeEnd}
          onPause={pause}
          onCanPlay={canPlay}
          onMute={mute}
          onFastBack={fastBack}
          onForward={forward}
        />
      </>
    )
  }
  const { getByText, container } = render(<TestComponent />)
  expect(container.querySelector('.nut-audio')).toHaveClass(
    'nut-audio audio-jest'
  )
  expect(container.querySelector('.nut-audio')).toHaveStyle('font-size: 20px;')
  const backButton = getByText('快退')
  const forwardButton = getByText('快进')
  const muteButton = getByText('静音')

  if (backButton && forwardButton && muteButton) {
    fireEvent.click(forwardButton)
    fireEvent.click(backButton)
    fireEvent.click(muteButton)
    expect(forward).toBeCalled()
    expect(fastBack).toBeCalled()
    expect(mute).toBeCalled()
  }
})

test('icon type test', () => {
  const { container } = render(<Audio type="icon" />)
  expect(container).toMatchSnapshot()
})

test('none type test', () => {
  const { container } = render(<Audio type="none" />)
  expect(container).toMatchSnapshot()
})
