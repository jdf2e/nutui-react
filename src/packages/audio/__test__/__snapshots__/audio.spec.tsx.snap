// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`icon type test 1`] = `
<div>
  <div
    class="nut-audio"
  >
    <div
      class="nut-audio-icon"
    >
      <div
        class="nut-audio-icon-box nut-audio-icon-stop"
      >
        <svg
          aria-labelledby="Service"
          class="nut-icon nut-icon-Service "
          role="presentation"
          viewBox="0 0 1024 1024"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1024 657.72H905.85V531.69c0-228.43-177.23-413.54-393.85-413.54S118.15 303.26 118.15 531.69v126.03H0V531.69C0 236.3 228.43 0 512 0s512 236.31 512 531.69zM196.92 866.46C86.64 866.46 0 779.81 0 669.54s86.65-196.92 196.92-196.93 196.92 86.65 196.93 196.93-86.65 196.92-196.93 196.92m0-275.69c-43.32 0-78.77 35.45-78.77 78.77s35.45 78.77 78.77 78.77 78.77-35.45 78.77-78.77-35.45-78.77-78.77-78.77m630.16 275.69c-110.28 0-196.92-86.65-196.93-196.92s86.65-196.92 196.93-196.93S1024 559.26 1024 669.54s-86.65 196.92-196.92 196.92m0-275.69c-43.32 0-78.77 35.45-78.77 78.77s35.45 78.77 78.77 78.77 78.77-35.45 78.77-78.77-35.45-78.77-78.77-78.77m-149.66 429.29c-23.63 0-47.26-15.75-55.14-43.32-7.88-31.51 7.88-63.02 39.38-74.83 102.4-31.51 189.05-102.4 236.31-196.93 15.75-27.57 51.2-39.38 78.77-23.63s39.38 51.2 23.63 78.77c-63.02 122.09-177.23 216.61-307.2 256-3.94 3.94-11.81 3.94-15.75 3.94"
            fill="currentColor"
            fill-opacity="0.9"
          />
        </svg>
      </div>
    </div>
    <audio
      class="audioMain"
      preload="auto"
      src=""
    >
      <track
        kind="captions"
      />
    </audio>
  </div>
</div>
`;

exports[`none type test 1`] = `
<div>
  <div
    class="nut-audio"
  >
    <div
      class="nut-audio-none-container"
    />
    <audio
      class="audioMain"
      preload="auto"
      src=""
    >
      <track
        kind="captions"
      />
    </audio>
  </div>
</div>
`;
